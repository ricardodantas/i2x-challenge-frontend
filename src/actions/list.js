import { createAction } from 'redux-actions'
import axios from 'axios'
import {isEmpty} from 'lodash'

export const listFetching = createAction('LIST_FETCHING')
export const listError    = createAction('LIST_ERROR')
export const listSuccess  = createAction('LIST_SUCCESS')

export async function fetchList() {
  const token = sessionStorage.getItem('token')

  try {
    return await axios.get('https://i2x-challenge.herokuapp.com/ai/recording/list/', {
        headers: {
          "content-type": "application/json",
          "Authorization": `JWT ${token}`
        }
      })
  } catch (e) {
    return e.message
  }

}

export function getList() {


  return async dispatch => {
    dispatch(listFetching())
    const response = await fetchList()
    const {data} = response

    if (!isEmpty(data) && !isEmpty(data.results)) {
      dispatch(listSuccess(data))
    } else {
      dispatch(listError({message: response}))
    }

  }
}
