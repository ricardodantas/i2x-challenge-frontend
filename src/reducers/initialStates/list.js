export const initialState = {
  isFetching: false,
  hasFailed: false,
  hasSuccess: false,
  items: [],
}
