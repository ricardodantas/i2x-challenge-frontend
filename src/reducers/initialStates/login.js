import {isEmpty} from 'lodash'
import apiAuth from '../../utils/auth'

// @flow
export const initialState: {
  isFetching: boolean,
  hasFailed: boolean,
  hasSuccess: boolean,
  response: Object<null>,
} = {
  isAuthenticated: apiAuth.isAuthenticated(),
  isFetching: false,
  hasFailed: false,
  hasSuccess: false,
  response: {},
}
