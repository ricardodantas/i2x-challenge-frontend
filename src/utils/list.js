// import FetchApi from './fetchApi'
import Auth from './auth'
import {isEmpty} from 'lodash'

export default class List extends Auth {

  static async getList() {
    try {
      return await super.get('ai/recording/list/', super.getToken())
    } catch (e) {
      throw new Error(e.message)
    }
  }
}
