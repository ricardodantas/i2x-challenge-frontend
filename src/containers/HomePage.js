import React, { Component } from 'react';
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import {getList} from '../actions/list'
import { Dimmer, Loader, Card, Rating } from 'semantic-ui-react'
import _ from 'lodash'
import moment from 'moment'
import ReactAudioPlayer from 'react-audio-player';

class HomePage extends Component {

  componentDidMount() {
    const {dispatch} = this.props
    dispatch(getList())
  }

  get items() {
    const {list} = this.props
    const {items} = list

    if(_.isEmpty(items)) {
      return null
    }

    return items.map((item) => {
      const duration =  parseInt(moment.duration(item.duration, 'seconds').asMinutes())
      const created = moment(item.created).from(new Date)

      return (
        <Card key={item.created}>
          <Card.Content>
            <Card.Header>
              <ReactAudioPlayer src={item.url} />
            </Card.Header>
            <Card.Meta>
              Duration: {duration} minutes | Created: {created}
            </Card.Meta>
            <Card.Description>
              {item.final_script}
            </Card.Description>
          </Card.Content>
          <Card.Content extra>
            <Rating icon='star' defaultRating={item.rating} maxRating={5} />
          </Card.Content>
        </Card>
      )
    })
  }

  render() {
    const {list} = this.props

    if(list.isFetching) {
      return (
        <Dimmer active>
          <Loader size="massive"/>
        </Dimmer>
      )
    }

    return (
      <div>
        <h3>Audio Recordings</h3>
        <Card.Group itemsPerRow={2}>
          {this.items}
        </Card.Group>
      </div>
    );
  }
}

const mapDispatchToProps = (dispatch) => {
  return {
    dispatch
  }
}

const mapStateToProps = (state) => {

  return {
    list: state.list
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(HomePage)
