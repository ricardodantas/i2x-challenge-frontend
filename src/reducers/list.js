import { handleActions } from 'redux-actions'
import {listFetching, listSuccess, listError} from '../actions/list'
import {initialState} from './initialStates/list'

export default handleActions({

  [listFetching]: (state, action) => {
    return Object.assign({}, state, {
      isFetching: true,
      hasFailed: false,
      hasSuccess: false,
      items: []
    })
  },
  [listSuccess]: (state, action) => {
    return Object.assign({}, state, {
      isFetching: false,
      hasFailed: false,
      hasSuccess: true,
      items: action.payload.results
    })
  },
  [listError]: (state, action) => {
    return Object.assign({}, state, {
      isFetching: false,
      hasFailed: true,
      hasSuccess: false,
      items: []
    })
  },
}, initialState)
