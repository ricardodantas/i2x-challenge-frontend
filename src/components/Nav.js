import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { Link, Redirect} from 'react-router-dom'
import { Container } from 'semantic-ui-react'
import {logout} from '../actions/login'

class Nav extends Component {

  static propTypes = {
    isAuthenticated: PropTypes.bool.isRequired,
    logout: PropTypes.func.isRequired,
  }

  logout() {
    const {dispatch, logout} = this.props

    dispatch(logout())
  }

  render() {
    const {isAuthenticated} = this.props

    if(!isAuthenticated) {
      return null
    }

    return (
      <div className="ui attached stackable menu">
        <Container>
          <Link to="/" className="active item">Home Page</Link>
          <div className="right menu">
            <a className="ui item" onClick={this.logout.bind(this)}>
              Logout
            </a>
          </div>
        </Container>
      </div>
    )
  }
}


const mapDispatchToProps = (dispatch) => {
  return {
    dispatch,
    logout
  }
}

const mapStateToProps = (state) => {
  return {}
}

export default connect(mapStateToProps, mapDispatchToProps)(Nav)
