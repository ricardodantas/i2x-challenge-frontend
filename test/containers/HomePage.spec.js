import React from 'react'
import { Provider } from 'react-redux'
import { mount } from 'enzyme'
import renderer from 'react-test-renderer'
import HomePage from '../../src/containers/HomePage'
import Nav from '../../src/components/Nav'
import configureStore from '../../src/store/configureStore'
import {initialState} from '../../src/reducers/initialStates/list'

describe('<HomePage />', () => {


  beforeEach(() => {
    global.store = configureStore({ list: initialState })
  })

  it('should display home page', () => {

    const tree = renderer.create(
      <Provider store={global.store}>
        <HomePage list={{}}/>
      </Provider>
    ).toJSON()

    expect(tree).toMatchSnapshot()
  })

})
