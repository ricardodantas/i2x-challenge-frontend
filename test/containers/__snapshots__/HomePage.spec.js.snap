// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HomePage /> should display home page 1`] = `
<div
  className="ui active transition visible dimmer"
  onClick={[Function]}
>
  <div
    className="content"
  >
    <div
      className="center"
    >
      <div
        className="ui massive loader"
      />
    </div>
  </div>
</div>
`;
